cpp/src/arrow/compute/api_scalar.cc

```diff
@@ -344,7 +367,8 @@ static auto kRoundToMultipleOptionsType = GetFunctionOptionsType<RoundToMultiple
     DataMember("round_mode", &RoundToMultipleOptions::round_mode));
 static auto kSetLookupOptionsType = GetFunctionOptionsType<SetLookupOptions>(
     DataMember("value_set", &SetLookupOptions::value_set),
-    DataMember("skip_nulls", &SetLookupOptions::skip_nulls));
+    DataMember("skip_nulls", &SetLookupOptions::skip_nulls),
+    DataMember("null_matching_behavior", &SetLookupOptions::null_matching_behavior));
```

Comment:
I think it'd be worthwhile to change this to only report/accept assignment to null_matching_behavior:
```suggestion
    CoercedDataMember("null_matching_behavior", &SetLookupOptions::null_matching_behavior,
      &SetLookupOptions::GetNullMatchingBehavior));
```

(This only affects how we deal with SetLookupOptions in reflected contexts like stringification; we'll still be able to assign to `skip_nulls`.) This can be enabled by putting the following into `reflection_internal.h`:

```c++
template <typename C, typename T>
struct CoercedDataMemberProperty {
  using Class = C;
  using Type = T;

  constexpr const Type& get(const Class& obj) const { return obj.*get_coerced_(); }

  void set(Class* obj, Type value) const { (*obj).*ptr_for_set_ = std::move(value); }

  constexpr std::string_view name() const { return name_; }

  std::string_view name_;
  Type Class::*ptr_for_set_;
  Type (Class::*get_coerced_)() const;
};

template <typename Class, typename Type>
constexpr CoercedDataMemberProperty<Class, Type> CoercedDataMember(std::string_view name,
                                                                   Type Class::*ptr,
                                                                   Type (Class::*get)()
                                                                       const) {
  return {name, ptr, get};
}
```

Reply:
![image](https://github.com/apache/arrow/assets/18380073/6ffb667a-863f-4a30-9f87-747d9b553b59)
Hi, here's an error I have no idea how to handle it when I put it to `reflection_internal.h`. Do you have any solution about it?

Reply:
It works!